src/plugins/intel_cpu/src/nodes/scaled_attn.cpp

```diff
@@ -193,17 +197,142 @@ struct MHAKernel<ScaledDotProductAttention::KT_ONEDNN, T> {
     // q: [B, H, q_len, S]
     // k: [B, H, kv_len, S]
     // v: [B, H, kv_len, S]
+    const GraphContext::CPtr context;
     dnnl::memory::desc q_md;
     dnnl::memory::desc k_md;
     dnnl::memory::desc weight_md;
     dnnl::memory::desc v_md;
     dnnl::memory::desc out_md;
     dnnl::memory attn_score;
     dnnl::memory attn_weight;
+    PlainTensor fp32_out;
+    PlainTensor qk_scratch_a;
+    PlainTensor qk_scratch_b;
+    PlainTensor wv_scratch_a;
+    PlainTensor wv_scratch_b;
+    std::vector<size_t> wsp;
+    size_t wsp_size_per_thread = 0;
     dnnl::matmul qk_prim;
     dnnl::matmul wv_prim;
     using tag = dnnl::memory::format_tag;
     using dt = dnnl::memory::data_type;
+    struct brgemmKey {
+        size_t M;
+        size_t N;
+        size_t K;
+        size_t lda;
+        size_t ldb;
+        size_t ldc;
+        bool b_transposed;
+        size_t hash() const {
+            using namespace dnnl::impl;
+            using namespace dnnl::impl::primitive_hashing;
+            size_t seed = 0;
+            seed = hash_combine(seed, M);
+            seed = hash_combine(seed, N);
+            seed = hash_combine(seed, K);
+            seed = hash_combine(seed, lda);
+            seed = hash_combine(seed, ldb);
+            seed = hash_combine(seed, ldc);
+            seed = hash_combine(seed, b_transposed);
+            return seed;
+        }
+        bool operator==(const brgemmKey& rhs) const {
+            return (rhs.M == M) && (rhs.N == N) && (rhs.K == K) && (rhs.lda == lda) && (rhs.ldb == ldb) &&
+                   (rhs.ldc == ldc) && (rhs.b_transposed == b_transposed);
+        }
+    };
+
+    std::shared_ptr<brgemmExecutor> qk_gemm_ptr = nullptr;
+    std::shared_ptr<brgemmExecutor> wv_gemm_ptr = nullptr;
+
+    MHAKernel() = delete;
+    explicit MHAKernel(GraphContext::CPtr ctx)
+        : context(ctx),
+          fp32_out(true),
+          qk_scratch_a(true),
+          qk_scratch_b(true),
+          wv_scratch_a(true),
+          wv_scratch_b(true) {}
+
+    void prepare_multiquery_prim(dnnl::stream strm,
+                                 PlainTensor& query,
+                                 PlainTensor& present_key,
+                                 bool has_out_transpose) {
+        auto make_dnnl_dims = [](const std::vector<size_t>& dims) {
+            dnnl::memory::dims dnnl_dims(dims.size());
+            for (size_t i = 0; i < dims.size(); i++)
+                dnnl_dims[i] = static_cast<dnnl::memory::dim>(dims[i]);
+            return dnnl_dims;
+        };
```

Comment:
How do you think to move this lambda function a private or a static helper function to remove the duplicate definition?

Reply:
That could be done.

Reply:
fix in 6ca10941f2955bb8be335626dcc9420a0af7faa5